File: lox-zig/src/scanner.zig
// zig fmt: off
const std = @import("std");

pub const Cursor = struct {
    source: []const u8,
    position: usize,
    prev: ?u8,
    curr: ?u8,

    pub fn peek(self: *const Cursor) ?u8 {
        return self.peek_nth(0);
    }

    pub fn peek_nth(self: *const Cursor, nth: usize) ?u8 {
        if (self.position + nth < self.source.len) {
            return self.source[self.position + nth];
        }

        return null;
    }

    pub fn bumb(self: *Cursor) void {
        if (self.position < self.source.len) {
            self.prev = self.curr;
            self.curr = self.source[self.position];
            self.position += 1;
        }
    }

    pub fn next(self: *Cursor) ?u8 {
        if (self.position < self.source.len) {
            self.prev = self.curr;
            self.curr = self.source[self.position];
            self.position += 1;

            return self.curr;
        }

        return null;
    }

    pub fn new(source: []const u8) Cursor {
        return .{
          .source = source,
          .position = 0,
          .prev = null,
          .curr= null
        };
    }
};

pub const Scanner = struct {
    cursor: Cursor,
    start: usize,

    pub fn new(source: []const u8) Scanner {
        return .{
            .cursor = Cursor.new(source),
            .start = 0,
        };
    }

    pub fn parse_space(self: *Scanner) TokenType {
        const EMPTY: []const u8  = " \t\r\n";
        while (self.cursor.peek()) |c_peek| {
            _ = std.mem.indexOfScalar(u8, EMPTY, c_peek) orelse break;

            self.cursor.bumb();
        }

        return TokenType.Whitespace;
    }

    pub fn on_match(s: *@This(), char: u8, tt: TokenType) ?TokenType {
        if (s.cursor.peek() orelse 0 == char) {
            s.cursor.bumb();
            return tt;
        } else {
            return null;
        }
    }

    fn try_parse_comment(s: *@This()) ?TokenType {
        if (s.cursor.peek() orelse 0 == '/') {
            while (s.cursor.peek() orelse '\n' != '\n') s.cursor.bumb();
            return TokenType.CommentLine;
        }

        return null;
    }

    fn try_parse_string(s: *@This()) ?TokenType {
        while (true) {
            s.cursor.bumb();
            if (s.cursor.peek() orelse { return null; } == '"') {
                s.cursor.bumb();
                return TokenType.String;  
            }
        }
    }

    fn bump_while(s: *@This(), f: fn(u8) bool) void {
        while (f(s.cursor.peek() orelse 0)) {
            s.cursor.bumb();
        }
    }

    fn try_parse_number(s: *@This()) ?TokenType {
        var seen_dot = false;
        while (true) switch (s.cursor.peek() orelse 0) {
            '0'...'9' => { s.cursor.bumb(); },
            '.' => switch (s.cursor.peek_nth(1) orelse 0) {
                '0'...'9' => {
                    s.cursor.bumb();
                    if (seen_dot) {
                        s.bump_while(struct { fn f(c: u8) bool {
                            return switch (c) {
                                '0'...'9' => true,
                                '.' => true,
                                else => false
                            };
                        } }.f);
                        return null;
                    }
                    seen_dot = true;
                },
                // Allow `9.9.sqrt()`
                //       `9.sqrt()`
                else => return .Number,
            },
            else => return .Number,
        };
    }

    fn prod_err(s: *@This(), kind: ErrorKind) Result {
        return Result{
            .Err = Error.new(kind, .{
                .start = s.start,
                .end = s.cursor.position
            })
        };
    }

    fn parse_reserved(s: *@This()) ?TokenType {
        s.bump_while(struct { fn f(c: u8) bool {
            return switch (c) {
                '0'...'9', 'a'...'z', 'A'...'Z', '_' => true,
                else => false,
            };
        }}.f);

        const string: []const u8 = s.cursor.source[s.start..s.cursor.position];

        switch (string.len) {
            2 => {
                if (std.mem.eql(u8, string, "if")) return .If;
                if (std.mem.eql(u8, string, "or")) return .Or;
            },
            3 => {
                if (std.mem.eql(u8, string, "and")) return .And;
                if (std.mem.eql(u8, string, "for")) return .For;
                if (std.mem.eql(u8, string, "fun")) return .Fun;
                if (std.mem.eql(u8, string, "var")) return .Var;
                if (std.mem.eql(u8, string, "nil")) return .Nil;
            },
            4 => {
                if (std.mem.eql(u8, string, "else")) return .Else;
                if (std.mem.eql(u8, string, "true")) return .True;
                if (std.mem.eql(u8, string, "this")) return .This;
            },
            5 => {
                if (std.mem.eql(u8, string, "class")) return .Class;
                if (std.mem.eql(u8, string, "false")) return .False;
                if (std.mem.eql(u8, string, "print")) return .Print;
                if (std.mem.eql(u8, string, "super")) return .Super;
                if (std.mem.eql(u8, string, "while")) return .While;
            },
            6 => {
                if (std.mem.eql(u8, string, "return")) return .Return;
            },
            else => return null,
        }
            
        return null;
    }

    pub fn next(s: *Scanner) ?Result {
        const c = s.cursor.next() orelse return null;
        s.start = s.cursor.position - 1;

        const res = switch (c) {
            'a'...'z', 'A'...'Z', '_' => s.parse_reserved() orelse TokenType.Identifier,
            ' ', '\t', '\r', '\n' => s.parse_space(),
            '0'...'9' => s.try_parse_number() orelse return s.prod_err(.InvalidNumber),
            '(' => .LeftParen,
            ')' => .RightParen,
            '{' => .LeftBrace,
            '}' => .RightBrace,
            ',' => .Comma,
            '.' => .Dot,
            '-' => .Minus,
            '+' => .Plus,
            ';' => .Semicolon,
            '*' => .Star,
            '!' => s.on_match('=', .BangEqual) orelse .Bang,
            '=' => s.on_match('=', .EqualEqual) orelse .Equal,
            '<' => s.on_match('=', .LessEqual) orelse .Less,
            '>' => s.on_match('=', .GreaterEqual) orelse .Greater,
            '/' => s.try_parse_comment() orelse .Slash,
            '"' => s.try_parse_string() orelse return s.prod_err(.UnfinishString),
            else => return s.prod_err(.Unknown)
        };

        return Result{.Ok = Token.new(res, .{
           .start = s.start,
           .end = s.cursor.position, 
        })};
    } 
};

pub const TokenType = enum(u8) {
    // Single-character tokens.
    LeftParen, RightParen, LeftBrace,
    RightBrace, Comma, Dot,
    Minus, Plus, Semicolon,
    Slash, Star,
    // One Or Two Character Tokens.
    //"!"
    Bang,
    CommentLine,
    // "! ="
    BangEqual, Equal, EqualEqual,
    Greater, GreaterEqual, Less,
    LessEqual,
    // Literals.
    Identifier, String,
    Number,
    // Keywords.
    And, Class,
    Else, False, Fun,
    For, If, Nil,
    Or,
    Print, Return,
    Super, This, True, Var, While,
    Eof, Whitespace
};

pub const ErrorKind = enum(u8){
    Unknown,
    UnfinishString,
    InvalidNumber,
};

pub const Error = struct {
   kind : ErrorKind,
   range : Range, 

   fn new(kind : ErrorKind, range : Range) Error{
       return.{
           .kind = kind,
           .range = range,
       };
   } 
};

pub const Result = union(enum){
    Ok:Token,
    Err:Error,
};

pub const Range = struct {
    start: usize,
    end: usize
};

pub const Token = struct {
    type: TokenType,
    range: Range,

    pub fn new(vtype: TokenType , range: Range) Token{
        return.{
           .type = vtype,
           .range = range
        };
    }
};

File: lox/src/scanner.rs
use crate::span::Span;
use std::ops::Not;

type Tk = TokenKind;

pub struct Scanner<'src> {
    cursor: Cursor<'src>,
    start: usize,
}

impl<'src> Scanner<'src> {
    pub fn new(src: &'src str) -> Scanner {
        Scanner {
            cursor: Cursor::new(src),
            start: 0,
        }
    }
}

#[derive(Debug)]
pub struct Error {
    pub span: Span,
    pub kind: ErrorKind,
}

impl Error {
    fn new(kind: ErrorKind, span: Span) -> Self {
        Error { span, kind }
    }
}

#[derive(Debug)]
pub enum ErrorKind {
    UnfinishedStr,
    UnknownToken,
    InvalidNumber,
}

impl Iterator for Scanner<'_> {
    type Item = Result<Token, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        let c = self.cursor.next()?;
        self.start = self.cursor.position - 1;

        match self.parse_next(c) {
            Ok(tt) => Some(Ok(Token::new(
                tt,
                Span::from(self.start..self.cursor.position),
            ))),
            Err(err) => Some(Err(Error::new(
                err,
                Span::from(self.start..self.cursor.position),
            ))),
        }
    }
}

impl<'src> Scanner<'src> {
    fn parse_next(&mut self, c: char) -> Result<TokenKind, ErrorKind> {
        Ok(match c {
            'a'..='z' | 'A'..='Z' | '_' => self.parse_reserved().unwrap_or(Tk::Identifier),
            '0'..='9' => self.parse_number().ok_or(ErrorKind::InvalidNumber)?,
            ' ' | '\n' | '\t' | '\r' => self.parse_space(),
            '(' => Tk::LeftParen,
            ')' => Tk::RightParen,
            '{' => Tk::LeftBrace,
            '}' => Tk::RightBrace,
            ',' => Tk::Comma,
            '.' => Tk::Dot,
            '-' => Tk::Minus,
            '+' => Tk::Plus,
            ';' => Tk::Semicolon,
            '*' => Tk::Star,
            '!' => self.on_match('=', |_| Tk::BangEqual).unwrap_or(Tk::Bang),
            '=' => self.on_match('=', |_| Tk::EqualEqual).unwrap_or(Tk::Equal),
            '>' => self
                .on_match('=', |_| Tk::GreaterEqual)
                .unwrap_or(Tk::Greater),
            '<' => self.on_match('=', |_| Tk::LessEqual).unwrap_or(Tk::Less),
            '/' => self
                .on_match('/', |s| {
                    while s.cursor.peek().unwrap_or('\n') != '\n' {
                        s.cursor.bump()
                    }

                    Tk::CommentLine
                })
                .unwrap_or(Tk::Slash),
            '"' => self.parse_string().ok_or(ErrorKind::UnfinishedStr)?,
            _ => return Err(ErrorKind::UnknownToken),
        })
    }
}

impl<'src> Scanner<'src> {
    fn on_match(
        &mut self,
        char: char,
        mut action: impl FnMut(&mut Self) -> TokenKind,
    ) -> Option<TokenKind> {
        matches!(self.cursor.peek(), Some(c) if c == char).then(|| {
            self.cursor.bump();
            action(self)
        })
    }
}

impl<'src> Scanner<'src> {
    fn parse_space(&mut self) -> TokenKind {
        let empty = [' ', '\t', '\r', '\n'];
        while let Some(c) = self.cursor.peek() {
            if empty.contains(&c) {
                self.cursor.bump();
            } else {
                break;
            }
        }

        TokenKind::Whitespace
    }

    fn bump_while(&mut self, predicate: impl Fn(char) -> bool) {
        while predicate(self.cursor.peek().unwrap_or_default()) {
            self.cursor.bump()
        }
    }

    fn parse_reserved(&mut self) -> Option<TokenKind> {
        self.bump_while(|c| c.is_ascii_digit() || c.is_ascii_alphabetic() || c == '_');
        Some(match &self.cursor.orig[self.start..self.cursor.position] {
            "if" => Tk::If,
            "or" => Tk::Or,
            "and" => Tk::And,
            "for" => Tk::For,
            "fun" => Tk::Fun,
            "var" => Tk::Var,
            "nil" => Tk::Nil,
            "else" => Tk::Else,
            "true" => Tk::True,
            "this" => Tk::This,
            "class" => Tk::Class,
            "false" => Tk::False,
            "print" => Tk::Print,
            "super" => Tk::Super,
            "while" => Tk::While,
            "return" => Tk::Return,
            _ => return None,
        })
    }

    fn parse_number(&mut self) -> Option<TokenKind> {
        let mut punto = false;

        while let Some(c) = self.cursor.peek() {
            let nxt_is_num = || matches!(self.cursor.peek_nth(1), Some('0'..='9'));
            match c {
                '0'..='9' => self.cursor.bump(),
                '.' if nxt_is_num() && punto => {
                    self.bump_while(|c| c.is_ascii_digit() || c == '.');
                    return None;
                }
                '.' if nxt_is_num() && !punto => {
                    self.cursor.bump();
                    punto = true
                }
                _ => break,
            }
        }

        Some(TokenKind::Number)
    }

    fn parse_string(&mut self) -> Option<TokenKind> {
        while let Some(c) = self.cursor.peek() {
            if c == '"' {
                self.cursor.bump();
                return Some(TokenKind::String);
            } else if ['\n', '\r'].contains(&c) {
                return None;
            } else {
                self.cursor.bump();
            }
        }

        None
    }
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy, Default)]
pub enum TokenKind {
    And,
    Bang,
    BangEqual,
    Class,
    Comma,
    CommentLine,
    Dot,
    #[default]
    Eof,
    Else,
    Equal,
    EqualEqual,
    False,
    For,
    Fun,
    Greater,
    GreaterEqual,
    If,
    Identifier,
    LeftBrace,
    LeftParen,
    Less,
    LessEqual,
    Minus,
    Nil,
    Number,
    Or,
    Print,
    Plus,
    Return,
    RightBrace,
    RightParen,
    Super,
    Semicolon,
    Slash,
    Star,
    String,
    This,
    True,
    Var,
    While,
    Whitespace,
}

#[derive(Debug, Clone, Copy)]
pub struct Token {
    pub tipo: TokenKind,
    pub span: Span,
}

impl Token {
    fn new(vtipo: TokenKind, span: Span) -> Self {
        Token { tipo: vtipo, span }
    }
}
struct Cursor<'src> {
    source: &'src str,
    orig: &'src str,
    prev: Option<char>,
    curr: Option<char>,
    position: usize,
}
impl<'src> Cursor<'src> {
    fn new(src: &'src str) -> Cursor {
        Cursor {
            source: src,
            orig: src,
            prev: None,
            curr: None,
            position: 0,
        }
    }

    fn peek(&self) -> Option<char> {
        self.peek_nth(0)
    }

    fn peek_nth(&self, nth: usize) -> Option<char> {
        self.source.chars().nth(nth)
    }

    fn bump(&mut self) {
        if self.source.is_empty().not() {
            self.prev = self.curr;
            self.curr = self.source.chars().next();
            self.source = &self.source[1..];
            self.position += 1;
        }
    }

    fn next(&mut self) -> Option<char> {
        self.prev = self.curr;
        match self.source.chars().next() {
            Some(c) => {
                self.curr = Some(c);
                self.source = &self.source[1..];
                self.position += 1;
                Some(c)
            }
            None => None,
        }
    }
}
